js/net.ts

```diff
@@ -0,0 +1,151 @@
+// Copyright 2018 the Deno authors. All rights reserved. MIT license.
+
+import { ReadResult, Reader, Writer, Closer } from "./io";
+import * as fbs from "gen/msg_generated";
+import { assert, notImplemented } from "./util";
+import * as dispatch from "./dispatch";
+import { flatbuffers } from "flatbuffers";
+import { read, write, close } from "./files";
+
+export type Network = "tcp";
+// TODO support other types:
+// export type Network = "tcp" | "tcp4" | "tcp6" | "unix" | "unixpacket";
+
+// TODO Support finding network from Addr, see https://golang.org/pkg/net/#Addr
+export type Addr = string;
+
+/** A Listener is a generic network listener for stream-oriented protocols. */
+export interface Listener {
+  /** accept() waits for and returns the next connection to the Listener. */
+  accept(): Promise<Conn>;
+
+  /** Close closes the listener.
+   * Any pending accept promises will be rejected with errors.
+   */
+  close(): void;
+
+  addr(): Addr;
+}
+
+class ListenerImpl implements Listener {
+  constructor(readonly fd: number) {}
+
+  async accept(): Promise<Conn> {
+    const builder = new flatbuffers.Builder();
+    fbs.Accept.startAccept(builder);
+    fbs.Accept.addRid(builder, this.fd);
+    const msg = fbs.Accept.endAccept(builder);
+    const baseRes = await dispatch.sendAsync(builder, fbs.Any.Accept, msg);
+    assert(baseRes != null);
+    assert(fbs.Any.NewConn === baseRes!.msgType());
+    const res = new fbs.NewConn();
+    assert(baseRes!.msg(res) != null);
+    return new ConnImpl(res.rid(), res.remoteAddr()!, res.localAddr()!);
+  }
+
+  close(): void {
+    close(this.fd);
+  }
+
+  addr(): Addr {
+    return notImplemented();
+  }
+}
+
+export interface Conn extends Reader, Writer, Closer {
+  localAddr: string;
+  remoteAddr: string;
+}
+
+class ConnImpl implements Conn {
+  constructor(
+    readonly fd: number,
+    readonly remoteAddr: string,
+    readonly localAddr: string
+  ) {}
+
+  write(p: ArrayBufferView): Promise<number> {
```

Comment:
I'm missing something like shutdown here. How is that done?

Reply:
It's not implemented yet. 
It will be added to the [AsyncWrite](https://docs.rs/tokio-io/0.1/tokio_io/trait.AsyncWrite.html) implementation in resources.rs.  There's a sub here
https://github.com/denoland/deno/blob/e5e7f0f038494bfe8aa14e31c8d13d6cf481186a/src/resources.rs#L104-L106

Reply:
https://golang.org/pkg/net/#TCPConn.CloseRead
Following golang I will surface it as `closeRead()`